platform/storage/posix/src/mender-storage.c

```diff
@@ -151,24 +151,101 @@ mender_storage_delete_authentication_keys(void) {
 }
 
 mender_err_t
-mender_storage_set_deployment_data(char *deployment_data) {
+mender_storage_set_deployment_data(cJSON *deployment_data) {
     assert(NULL != deployment_data);
-    size_t deployment_data_length = strlen(deployment_data);
 
-    if (MENDER_OK != mender_storage_write_file(MENDER_STORAGE_NVS_DEPLOYMENT_DATA, deployment_data, deployment_data_length)) {
+    /* Add/update version number. We always want to store the deployment data in
+     * the format of the latest and greatest version. By updating the version
+     * number here and validating the data below, we make sure to enforce this.
+     */
+    cJSON *item;
+    if (NULL == (item = cJSON_CreateNumber((double)MENDER_STORAGE_DEPLOYMENT_DATA_VERSION))) {
+        mender_log_error("Unable to allocate memory");
+        return MENDER_FAIL;
+    }
+    if (NULL == cJSON_GetObjectItem(deployment_data, "version")) {
+        if (!cJSON_AddItemToObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    } else {
+        /* One would expect cJSON_AddItemToObject above to replace the value on
+         * already existing keys. However, it creates another entry with a
+         * duplicate key, which is invalid JSON. So we have to be careful. */
+        if (!cJSON_ReplaceItemInObject(deployment_data, "version", item)) {
+            mender_log_error("Unable to allocate memory");
+            cJSON_Delete(item);
+            return MENDER_FAIL;
+        }
+    }
```

Comment:
An alternative would be to always `cJSON_DeleteItemFromObject()` (which does nothing in case it's not there) and then add it.